File: llm/lib/prompts.mjs
// Prompt file resolution and seeding.
// Defaults live in templates/prompts/; editable per-KB copies live in .kb/config/prompts/.
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = path.resolve(__dirname, '..', '..');

export function defaultPromptsDir() {
  return path.join(REPO_ROOT, 'templates', 'prompts');
}

export function kbPromptsDir(kbRoot) {
  return path.join(kbRoot, '.kb', 'config', 'prompts');
}

export function resolvePrompt(kbRoot, name) {
  const kbPath = path.join(kbPromptsDir(kbRoot), `${name}.md`);
  if (fs.existsSync(kbPath)) return fs.readFileSync(kbPath, 'utf8');
  const defaultPath = path.join(defaultPromptsDir(), `${name}.md`);
  if (fs.existsSync(defaultPath)) return fs.readFileSync(defaultPath, 'utf8');
  throw new Error(`prompt not found: ${name}`);
}

export function listPrompts() {
  const dir = defaultPromptsDir();
  if (!fs.existsSync(dir)) return [];
  return fs.readdirSync(dir)
    .filter((f) => f.endsWith('.md'))
    .map((f) => f.slice(0, -3));
}

export function initPrompts(kbRoot, { force = false } = {}) {
  const src = defaultPromptsDir();
  const dst = kbPromptsDir(kbRoot);
  fs.mkdirSync(dst, { recursive: true });

  if (!fs.existsSync(src)) {
    throw new Error('default prompts directory missing: ' + src);
  }

  const results = [];
  for (const file of fs.readdirSync(src)) {
    if (!file.endsWith('.md')) continue;
    const srcPath = path.join(src, file);
    const dstPath = path.join(dst, file);
    if (fs.existsSync(dstPath) && !force) {
      results.push({ file, status: 'skipped' });
      continue;
    }
    fs.copyFileSync(srcPath, dstPath);
    results.push({ file, status: force && fs.existsSync(dstPath) ? 'overwritten' : 'created' });
  }
  return { src, dst, results };
}

File: agent/weft_agent/prompts.py
"""Prompt file resolution and seeding (port of llm/lib/prompts.mjs).
Defaults live in <repo>/templates/prompts/; editable per-KB copies in .kb/config/prompts/.
"""
import functools
import shutil
from pathlib import Path

from .textutil import REPO_ROOT


def default_prompts_dir() -> Path:
    return REPO_ROOT / "templates" / "prompts"


def kb_prompts_dir(kb_root: Path) -> Path:
    return kb_root / ".kb" / "config" / "prompts"


@functools.lru_cache(maxsize=None)
def resolve_prompt(kb_root: Path, name: str) -> str:
    # cached per process (init-prompts clears after writing)
    kb_path = kb_prompts_dir(kb_root) / f"{name}.md"
    if kb_path.exists():
        return kb_path.read_text(encoding="utf-8")
    default_path = default_prompts_dir() / f"{name}.md"
    if default_path.exists():
        return default_path.read_text(encoding="utf-8")
    raise FileNotFoundError(f"prompt not found: {name}")


def list_prompts() -> list[str]:
    d = default_prompts_dir()
    if not d.is_dir():
        return []
    return sorted(f.stem for f in d.glob("*.md"))


def init_prompts(kb_root: Path, force: bool = False) -> dict:
    src = default_prompts_dir()
    dst = kb_prompts_dir(kb_root)
    dst.mkdir(parents=True, exist_ok=True)
    if not src.is_dir():
        raise FileNotFoundError(f"default prompts directory missing: {src}")

    results = []
    for src_path in sorted(src.glob("*.md")):
        dst_path = dst / src_path.name
        if dst_path.exists() and not force:
            results.append({"file": src_path.name, "status": "skipped"})
            continue
        existed = dst_path.exists()
        shutil.copyfile(src_path, dst_path)
        results.append({"file": src_path.name, "status": "overwritten" if existed else "created"})
    return {"src": str(src), "dst": str(dst), "results": results}
